src/components/lib/sl/sl.h

```diff
@@ -395,33 +404,123 @@ sl_thd_is_runnable(struct sl_thd *t)
 	return (t->state == SL_THD_RUNNABLE || t->state == SL_THD_WOKEN);
 }
 
+int sl_thd_kern_dispatch(thdcap_t t);
+
 static inline int
-sl_thd_activate(struct sl_thd *t, sched_tok_t tok)
+sl_thd_activate(struct sl_thd *t, sched_tok_t tok, tcap_time_t timeout)
 {
 	struct cos_defcompinfo *dci = cos_defcompinfo_curr_get();
 	struct cos_compinfo    *ci  = &dci->ci;
-	struct sl_global_cpu   *g   = sl__globals_cpu();
+	struct sl_global_cpu  *g   = sl__globals_cpu();
 	int ret = 0;
 
 	if (t->properties & SL_THD_PROPERTY_SEND) {
-		return cos_sched_asnd(t->sndcap, g->timeout_next, g->sched_rcv, tok);
+		return cos_sched_asnd(t->sndcap, timeout, g->sched_rcv, tok);
 	} else if (t->properties & SL_THD_PROPERTY_OWN_TCAP) {
 		return cos_switch(sl_thd_thdcap(t), sl_thd_tcap(t), t->prio,
-				  g->timeout_next, g->sched_rcv, tok);
+				  timeout, g->sched_rcv, tok);
 	} else {
 		ret = cos_defswitch(sl_thd_thdcap(t), t->prio, t == g->sched_thd ?
-				    TCAP_TIME_NIL : g->timeout_next, tok);
+				    TCAP_TIME_NIL : timeout, tok);
 		if (likely(t != g->sched_thd && t != g->idle_thd)) return ret;
 		if (unlikely(ret != -EPERM)) return ret;
 
 		/*
 		 * Attempting to activate scheduler thread or idle thread failed for no budget in it's tcap.
 		 * Force switch to the scheduler with current tcap.
 		 */
-		return cos_switch(sl_thd_thdcap(g->sched_thd), 0, t->prio, 0, g->sched_rcv, tok);
+		return cos_switch(g->sched_thdcap, g->sched_tcap, t->prio,
+				  timeout, g->sched_rcv, tok);
 	}
 }
 
+static inline int
```

Comment:
Yeah, this is the core of the logic for slite. This is very very delicate, and we'll want to try and change it as little as possible. Phani took a lot of time on this, and I completely trust his judgement for where it ended up.

Reply:
...of course, we'll need to update for x86-64, and provide a default version that always uses the slowpath for architectures for which we don't have a fastpath.